docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,109 @@
+
+The benefit of the above is that the opinionated flow can be accessed in a more nuanced and piecemeal way while also gaining the potential for more advanced features. Say a consumer wants to:
+
+1. setup a custom config file for ilab (optional)
+2. initialize a taxonomy
+3. ensure their taxonomy is valid
+4. ingest some data for RAG and SDG (SDG coming soon)
+5. generate synthetic data using an InstructLab pipeline
+6. do some custom handling per their use case
+7. fine-tune a model using the custom config they initialized for their hardware
+
+a user could do the following to accomplish this if they had an SDK:
+
+```python
+instructlab.api.configuration.init(taxonomy=path_to_taxonomy)
+if instructlab.api.taxonomy.diff():
+    instructlab.api.rag.ingest(...)
+    instructlab.api.data.ingest(...)
+    data_paths = instructlab.api.data.generate(...)
+    some_custom_handling(data_paths)
+    instructlab.api.model.train(data_path=..., strategy=lab_multiphase)
+```
+
+However today, users are forced to run a sequence of commands tailored to only work with the proper directory structure on the system.
+
+## Major Goals
+
+1. Modularize the InstructLab workflow such that any part can be run independently
+2. Allow users to choose whether or not to take advantage of the config/system-profile method of running InstructLab. Meaning they do not need any pre-existing configuration to run the SDK.
+3. Standardize user contracts for the existing functionality of the InstructLab workflow. Existing CLI commands should be using the SDK once past click parsing, not separate code.
+4. Define Contracts loose enough that functionality can be expanded as more advanced features are released.
+5. Document SDK usage in upcoming InstructLab releases.
+
+## Non-Goals
+
+1. Exposing all library functionality immediately
+2. Replacing CLI
+3. Shipping an SDK that is generally available as opposed to v1alpha1 or v1beta1.
+
+## Design
+
+### Versioning
+
+The SDK would start at version v1alpha1 such that it can change/break at any time for the first few iterations as libraries adjust their API surface.
+
+### Structure
+
+This SDK should live in a net new package inside of `instructlab/instructlab` preferably to limit unnecessary imports in a new repository. It could be imported as `instructlab.api...`
+
+The user surface initially should look like this:
+
+```console
+instructlab.api.configuration.init
+instructlab.api.taxonomy.diff
```

Comment:
As this structure is what the current CLI has, I may be asking a larger question and won't be offended if you shut this as resolved to handle it somewhere else. Would it make more sense for the `.taxonomy.diff` tree to be under `.data`, to eventually leave room for alternative forms of data sorting and ingestion into the pipeline (i.e., is taxonomy a subfield of data management  overall, with SDG as another subfield)?

Reply:
new update (sorry), Taxonomy is going to get its own class after speaking with folks in the PR here and offline